### PR TITLE
[sw] Move our memory operation functions into lib/base.

### PR DESCRIPTION
**Pull request by mcy**
This change also introduces read_32() and write_32(), for doing direct word manipulation without violating strict alising. These functions will be useful for DIF implementations.

### CODE REVIEW

**Comment by silvestrst** [comment]
LGTM, but it is probably @lenary  to have a look

File: sw/device/lib/base/memory.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+ * Of course, |ptr| must point to word-aligned memory that is at least one word
+ * wide. To do otherwise is Undefined Behavior. Note that this memory need not
+ * be initialized.
+ *
+ * @param ptr a word-aligned pointer pointed to at least four bytes of memory.
+ * @return the word |ptr| points to.
+ */
+inline uint32_t read_32(void *ptr) {
+  uint32_t out;
+  asm volatile("lw %0, 0(%1)" : "=r"(out) : "r"(ptr));
```
Comment (lenary) [inline sw/device/lib/base/memory.h]: I don't think we need to do this, honestly.

You should be able to cast a `void*` to a `uint32_t*`, which is valid as long as the `void*` is aligned, which this function requires anyway. This has the advantage of totally avoiding inline (volatile) assembly.

The rules of aliasing don't really apply, as `void*` is assignment compatible with any other pointer type - C just makes it UB to dereference an underaligned pointer, a null pointer, or a few other kinds of pointer that don't apply in this case.

File: sw/device/lib/base/memory.c [hunk 2]
```diff
@@ -0,0 +1,35 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/memory.h"
+
+extern inline uint32_t read_32(void *);
```
Comment (lenary) [inline sw/device/lib/base/memory.c]: These should just be `extern`. I got this wrong in the first version of the C11 style guide and updated it.

File: sw/device/lib/base/memory.h [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+ * well-defined.
+ *
+ * Of course, |ptr| must point to word-aligned memory that is at least one word
+ * wide. To do otherwise is Undefined Behavior.
+ *
+ * @param value the value to store.
+ * @param ptr a word-aligned pointer pointed to at least four bytes of memory.
+ */
+inline void write_32(uint32_t value, void *ptr) {
+  asm volatile("sw %0, 0(%1)" ::"r"(value), "r"(ptr));
```
Comment (lenary) [inline sw/device/lib/base/memory.h]: Like `read_32`, this does not need to use inline assembly, just use a cast.

File: sw/device/lib/base/memory.h (hunk 1 above)
Comment (mcy) [reply sw/device/lib/base/memory.h]: I don't think that's true, unfortunately. My read of sections 6.5.5 and 6.5.6 is that if you have a `void*`, and you do not know a priori its effective type (per S6.5.5), the only lvalue you're guaranteed to know is ok is `char*`. In particular, my read of strict aliasing is that this is UB:
```
char bytes[4] alignas(4) = "abcd";
// ... across a function boundary ...
uint32_t word = *((uint32_t *)bytes);
```

Mind, normally I wouldn't care about function boundaries, but the explicit inline-ability here makes me nervous.

File: sw/device/lib/base/memory.c [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  uint8_t *dest8 = (uint8_t *)dest;
+  uint8_t value8 = (uint8_t)value;
+  for (size_t i = 0; i < len; ++i) {
+    dest8[i] = value8;
+  }
+  return dest;
+}
+
+noreturn void abort(void) {
+  asm volatile("unimp");
```
[the remark this answers is not in the document]
Comment (mcy) [reply sw/device/lib/base/memory.c]: I didn't know it had changed... oops!

**Comment by lenary** [comment]
@mcy and I have had a long discussion about the assembly snippets in `read_32` and `write_32`, which we both want to avoid. I was incorrect in thinking we can just use a cast.

We think we have a solution: Using `__builtin_assume_aligned`, as supported by both GCC and Clang (because the function already documents that it assumes the pointer is correctly aligned), and `__builtin_memcpy` to do the read/write itself. Using `__builtin_memcpy` works here because we know there's an implementation of `memcpy` to fall back on (not that it will need it, as the size operand will be a constant expression, and small enough that it should just turn into a single `lw` or `sw`), and `-fno-builtins` does not turn off `__builtin_*` functions, it only turns off recognising the unprefixed versions of them.

We believe this approach will satisfy any issues brought up by inlining the functions, as suggested.

@mcy please do correct me if I'm wrong about any parts of the above. 

I think this "base" library is exactly the place for this kind of careful implementation, rather than having these details spread around the OT software codebase.

**Comment by mcy** [comment]
@lenary thank you for making me look for a better alternative. I believe your comment is correct: strict aliasing forbids accepting a `void*`, even a well-aligned one, and using it to construct an lvalue of non-character type.

**Comment by mcy** [comment]
This is slightly off-topic, but given that Ibex does implement unaligned reads and writes, and that we have *some* consensus that weirdness is permissible in libbase, do we want to think about making memcpy/memset take advantage of this (since we could write a "vectorized" memcpy that uses sw/lw where possible)? Presumably that would be faster than whatever LLVM uses as a memcpy.

This probably doesn't matter though? Especially if LLVM gains knowledge of Ibex and its unaligned accesses. Either way, explicitly banning "we assume Ibex" is an option to consider as well.

**Comment by lenary** [comment]
Honestly, we shouldn't really be thinking about "faster" yet. I don't think we're getting held up by how slow our memcpy is.

Eventually there will be time for that kind of optimisation, but I don't think it should be on the agenda at the moment.

**Comment by silvestrst** [comment]
LGTM, thanks @mcy and @lenary

**Comment by asb** [comment]
Thanks for starting this cleanup, a couple of comments/questions: Why does `abort` live in memory.h? Also, this PR seems to remove strlen yet doesn't document that fact in the commit message. This makes me wonder whether it is intentional, or a mistake?

**Comment by mcy** [comment]
@asb: Abort is in there for now because:
- I want to move it into lib/base sooner rather than later.
- But, I don't want to decide a place for it yet. If you want me to pick a header for abort to live in, I'll do it (not a fan of "abort.h", but maybe something along the lines of "execution_control.h"?)

Also, I removed strlen() because in an earlier All OT SW meeting we decided we wanted to get rid of anything involving NUL-termination. strlen() is only used in one place, so I got rid of it. I can note this in the commit message, too.

**Comment by asb** [comment]
Thanks for the explanation @mcy, that's really helpful - all very sensible, just useful to have that background.

I think this is basically good to go but would like to see the following simple changes:
* Make the removal of strlen a separate commit. In that commit message, reference the fact we decided we didn't want standard lib functions involving NUL-termination
* `execution_control.h` sounds fine to me. If there's arguments about the header name, I wouldn't be opposed to have abort temporarily in memory.h as long as the fact it will probably move was mentioned in the commit message and probably and in source comment

Thanks!

**Comment by mcy** [comment]
I don't actually like `execution_control.h` that much, I was just throwing something out there.

I just want to pick a header name that encompasses all functions we'd like to group together. E.g., imagine we introduce our own `unreachable()`... deferring to a future change might not be the worst idea?

**Comment by lenary** [comment]
Let's defer to a future change. This question isn't orthogonal to issues around the freestanding headers, and providing additional C library symbols. I will start a google doc to discuss the issues around that stuff as I see them, but I want us to land this and the freestanding headers first.

**Comment by asb** [comment]
I'm happy to leave splitting out abort to a different file for a future change, but think we should have a comment in tree mentioning this, and likely an issue tracking it.

**Pull request merged**